src/Microsoft.FeatureManagement/FeatureManager.cs

```diff
@@ -217,6 +263,216 @@ await contextualFilter.EvaluateAsync(context, appContext).ConfigureAwait(false)
+            }
+
+            if (featureDefinition.Allocation == null)
+            {
+                throw new FeatureManagementException(
+                    FeatureManagementError.MissingAllocation,
+                    $"No allocation is defined for the feature {featureDefinition.Name}");
+            }
+
+            if (!featureDefinition.Variants?.Any() ?? false)
+            {
+                throw new FeatureManagementException(
+                    FeatureManagementError.MissingVariantDefinition,
+                    $"No variants are registered for the feature {feature}");
+            }
+
+            VariantDefinition variantDefinition;
+
+            bool isFeatureEnabled = await IsEnabledAsync(feature, context, true, cancellationToken).ConfigureAwait(false);
+
+            variantDefinition = await GetVariantDefinitionAsync(featureDefinition, context, useContext, isFeatureEnabled, cancellationToken).ConfigureAwait(false);
+
+            if (variantDefinition == null)
+            {
+                return null;
+            }
+
+            IConfigurationSection variantConfiguration = null;
+
+            bool configValueSet = !string.IsNullOrEmpty(variantDefinition.ConfigurationValue);
+            bool configReferenceValueSet = !string.IsNullOrEmpty(variantDefinition.ConfigurationReference);
+
+            if (configValueSet && configReferenceValueSet)
+            {
+                throw new FeatureManagementException(
+                    FeatureManagementError.InvalidVariantConfiguration,
+                    $"Both ConfigurationValue and ConfigurationReference are specified for the variant {variantDefinition.Name} in feature {feature}");
+            }
+            else if (configReferenceValueSet)
+            {
+                if (_configuration == null)
+                {
+                    throw new FeatureManagementException(
+                        FeatureManagementError.InvalidVariantConfiguration,
+                        $"Cannot use {nameof(variantDefinition.ConfigurationReference)} if no instance of {nameof(IConfiguration)} is present.");
+                }
+
+                variantConfiguration = _configuration.GetSection(variantDefinition.ConfigurationReference);
+            }
+            else if (configValueSet)
+            {
+                VariantConfigurationSection section = new VariantConfigurationSection(variantDefinition.Name, "", variantDefinition.ConfigurationValue);
+                variantConfiguration = section;
+            }
+
+            Variant returnVariant = new Variant()
+            {
+                Name = variantDefinition.Name,
+                Configuration = variantConfiguration
+            };
+
+            return returnVariant;
+        }
+
+        private async ValueTask<VariantDefinition> GetVariantDefinitionAsync(FeatureDefinition featureDefinition, TargetingContext context, bool useContext, bool isFeatureEnabled, CancellationToken cancellationToken)
+        {
```

Comment:
I don't expect a method like GetVariantDefinitionAsync should be resolving targeting context. That seems like a separate concern.

Reply:
I could move this logic to `AssignVariantAsync` instead